sdk/containerregistry/container-registry/src/blob/containerRegistryBlobClient.ts

```diff
@@ -0,0 +1,400 @@
+import { createSerializer, FullOperationResponse } from "@azure/core-client";
+import { readStreamToEnd } from "../utils/helpers";
+import { Readable } from "stream";
+import { createSpan } from "../tracing";
+import { SpanStatusCode } from "@azure/core-tracing";
+
+const LATEST_API_VERSION = "2021-07-01";
+
+enum KnownManifestMediaType {
+  OciManifestMediaType = "application/vnd.oci.image.manifest.v1+json",
+}
+
+function isReadableStream(body: any): body is NodeJS.ReadableStream {
+  return body && typeof body.pipe === "function";
+}
+
+const serializer = createSerializer(Mappers, /* isXML */ false);
+
+/**
+ * The Azure Container Registry blob client, responsible for uploading and downloading blobs and manifests, the building blocks of artifacts.
+ */
+export class ContainerRegistryBlobClient {
+  /**
+   * The Azure Container Registry endpoint.
+   */
+  public readonly endpoint: string;
+
+  /**
+   * The name of the repository that logically groups the artifact parts.
+   */
+  public readonly repositoryName: string;
+
+  private client: GeneratedClient;
+
+  /**
+   * Creates an instance of a ContainerRegistryBlobClient for managing container images and artifacts.
+   *
+   * Example usage:
+   * ```ts
+   * import { ContainerRegistryBlobClient } from "@azure/container-registry";
+   * import { DefaultAzureCredential} from "@azure/identity";
+   *
+   * const client = new ContainerRegistryBlobClient(
+   *    "<container registry API endpoint>",
+   *    "<repository name>",
+   *    new DefaultAzureCredential()
+   * );
+   * ```
+   * @param endpoint - the URL endpoint of the container registry
+   * @param repositoryName - the name of the repository that logically groups the artifact parts
+   * @param credential - used to authenticate requests to the service
+   * @param options - optional configuration used to send requests to the service
+   */
+  constructor(
+    endpoint: string,
+    repositoryName: string,
+    credential: TokenCredential,
+    options: ContainerRegistryClientOptions = {}
+  ) {
+    if (!endpoint) {
+      throw new Error("invalid endpoint");
+    }
+
+    this.endpoint = endpoint;
+    this.repositoryName = repositoryName;
+
+    const internalPipelineOptions: InternalPipelineOptions = {
+      ...options,
+      loggingOptions: {
+        logger: logger.info,
+        // This array contains header names we want to log that are not already
+        // included as safe. Unknown/unsafe headers are logged as "<REDACTED>".
+        additionalAllowedQueryParameters: [
+          "last",
+          "n",
+          "orderby",
+          "digest",
+          "_nouploadcache",
+          "_state",
+        ],
+      },
+    };
+    // Require audience now until we have a default ACR audience from the service.
+    if (!options.audience) {
+      throw new Error(
+        "ContainerRegistryClientOptions.audience must be set to initialize ContainerRegistryClient."
+      );
```

Comment:
this feels like bad typing - the options for this class aren't optional, and there is a required option in them. We should either declare an options interface that required audience and not have a default value for options in the constructor, or instead pull audience out into a required parameter after credential

either way, I don't think we should be re-using the options interface from the other client

Reply:
This check is copied verbatim from the other client, so I'd want to understand the original reasoning behind this before changing it (@jeremymeng do you know?). But can definitely add a separate options interface.

Reply:
It was decided that we would require users to pass in the audience at the moment. We can then make them optional after the service team implemented the feature to use a single audience for all clouds like storage has.  Making them required now but optional later would not be a breaking change.  We did this for all the languages.

Reply:
Got it, but was there any reason not to make the audience property required in the options type itself instead of having this runtime check?

Reply:
Mostly it's nice to give folks squiggles in their editor / errors when they compile during development rather than hitting an issue during runtime, we tend to not waste cycles on runtime checks since it slows down execution for correct calls while being harder to diagnose and discover than compiler errors

Reply:
Oh yeah I totally agree but was wondering why it _wasn't_ done like that in the original ContainerRegistryClient. It's not super important, more out of curiosity.

Reply:
I don't remember exactly, most likely I was mirroring what .NET did.

Reply:
Gotcha, I'll just make it required here then. Maybe we should consider changing it in the original client as well at some point? But I wonder if that could be a breaking change in some obscure scenario.

Reply:
If it has always thrown at runtime (e.g. nobody has ever been able to actually pass it optionally) I don't consider fixing the types to be more accurate to be a break